File: src/bmsx_cpp/render/shared/queues.cpp
/*
 * queues.cpp - Render submission queues implementation
 */

#include "queues.h"
#include "common/clamp.h"
#include <array>
#include <algorithm>

namespace bmsx {
namespace RenderQueues {

namespace {
enum class QueueSource : u8 {
	Front = 0,
	Back = 1,
};

FeatureQueue<MeshRenderSubmission> s_meshQueue(256);
FeatureQueue<ParticleRenderSubmission> s_particleQueue(1024);
QueueSource s_activeQueueSource = QueueSource::Front;

struct RenderQueueLifecycle {
	bool (*hasFront)();
	bool (*hasBack)();
	void (*swap)();
	void (*clearBack)();
	void (*clearAll)();
};

bool meshHasFront() { return s_meshQueue.sizeFront() > 0; }
bool meshHasBack() { return s_meshQueue.sizeBack() > 0; }
void meshSwap() { s_meshQueue.swap(); }
void meshClearBack() { s_meshQueue.clearBack(); }
void meshClearAll() { s_meshQueue.clearAll(); }

bool particleHasFront() { return s_particleQueue.sizeFront() > 0; }
bool particleHasBack() { return s_particleQueue.sizeBack() > 0; }
void particleSwap() { s_particleQueue.swap(); }
void particleClearBack() { s_particleQueue.clearBack(); }
void particleClearAll() { s_particleQueue.clearAll(); }

const std::array<RenderQueueLifecycle, 2> s_renderQueueLifecycles{{
	{meshHasFront, meshHasBack, meshSwap, meshClearBack, meshClearAll},
	{particleHasFront, particleHasBack, particleSwap, particleClearBack, particleClearAll},
}};

template<typename T>
const ScratchBatch<T>& activeQueue(FeatureQueue<T>& queue) {
	return s_activeQueueSource == QueueSource::Back ? queue.back() : queue.front();
}

} // namespace

i32 particleAmbientModeDefault = 0;
f32 particleAmbientFactorDefault = 1.0f;
std::array<f32, 3> _skyTint = {1.0f, 1.0f, 1.0f};
f32 _skyExposure = 1.0f;

static bool hasCommittedFrontQueueContent() {
	for (const RenderQueueLifecycle& lifecycle : s_renderQueueLifecycles) {
		if (lifecycle.hasFront()) return true;
	}
	return false;
}

void prepareCompletedRenderQueues() {
	for (const RenderQueueLifecycle& lifecycle : s_renderQueueLifecycles) {
		lifecycle.swap();
	}
	s_activeQueueSource = QueueSource::Front;
}

void preparePartialRenderQueues() {
	s_activeQueueSource = hasCommittedFrontQueueContent()
		? QueueSource::Front
		: (hasPendingBackQueueContent() ? QueueSource::Back : QueueSource::Front);
}

void prepareOverlayRenderQueues() {
	s_activeQueueSource = QueueSource::Back;
}

void prepareHeldRenderQueues() {
	s_activeQueueSource = QueueSource::Front;
}

bool hasPendingBackQueueContent() {
	for (const RenderQueueLifecycle& lifecycle : s_renderQueueLifecycles) {
		if (lifecycle.hasBack()) return true;
	}
	return false;
}

void clearBackQueues() {
	for (const RenderQueueLifecycle& lifecycle : s_renderQueueLifecycles) {
		lifecycle.clearBack();
	}
	s_activeQueueSource = QueueSource::Front;
}

void clearAllQueues() {
	for (const RenderQueueLifecycle& lifecycle : s_renderQueueLifecycles) {
		lifecycle.clearAll();
	}
	s_activeQueueSource = QueueSource::Front;
}

void submit_mesh(const MeshRenderSubmission& item) {
	s_meshQueue.submit(item);
}

i32 beginMeshQueue() {
	return static_cast<i32>(s_activeQueueSource == QueueSource::Back ? s_meshQueue.sizeBack() : s_meshQueue.sizeFront());
}

const MeshRenderSubmission& meshQueueEntry(size_t index) {
	return activeQueue(s_meshQueue).get(index);
}


void submit_particle(const ParticleRenderSubmission& item) {
	s_particleQueue.submit(item);
}

i32 beginParticleQueue() {
	return static_cast<i32>(s_activeQueueSource == QueueSource::Back ? s_particleQueue.sizeBack() : s_particleQueue.sizeFront());
}

const ParticleRenderSubmission& particleQueueEntry(size_t index) {
	return activeQueue(s_particleQueue).get(index);
}


void setAmbientDefaults(i32 mode, f32 factor) {
	particleAmbientModeDefault = mode;
	particleAmbientFactorDefault = clamp(factor, 0.0f, 1.0f);
}

void setSkyboxTintExposure(const std::array<f32, 3>& tint, f32 exposure) {
	_skyTint = {std::max(0.0f, tint[0]), std::max(0.0f, tint[1]), std::max(0.0f, tint[2])};
	_skyExposure = std::max(0.0f, exposure);
}

} // namespace RenderQueues
} // namespace bmsx

File: src/bmsx/render/shared/queues.ts
import { FeatureQueue } from '../../common/feature_queue';
import type {
	GlyphRenderSubmission,
	HostImageRenderSubmission,
	MeshRenderSubmission,
	ParticleRenderSubmission,
	PolyRenderSubmission,
	RectRenderSubmission,
} from './submissions';
export type Host2DKind = 'img' | 'poly' | 'rect' | 'glyphs';
export type Host2DRef = HostImageRenderSubmission | PolyRenderSubmission | RectRenderSubmission | GlyphRenderSubmission;
export type Host2DSubmission =
	| ({ type: 'img' } & HostImageRenderSubmission)
	| ({ type: 'poly' } & PolyRenderSubmission)
	| ({ type: 'rect' } & RectRenderSubmission)
	| ({ type: 'glyphs' } & GlyphRenderSubmission);

const meshQueue = new FeatureQueue<MeshRenderSubmission>(256);
const particleQueue = new FeatureQueue<ParticleRenderSubmission>(1024);
let activeQueueSource: 'front' | 'back' = 'front';

type RenderQueueLifecycle = {
	hasFront(): boolean;
	hasBack(): boolean;
	swap(): void;
	clearBack(): void;
	clearAll(): void;
};

const renderQueueLifecycles: readonly RenderQueueLifecycle[] = [
	{
		hasFront: () => meshQueue.sizeFront() > 0,
		hasBack: () => meshQueue.sizeBack() > 0,
		swap: () => { meshQueue.swap(); },
		clearBack: () => { meshQueue.clearBack(); },
		clearAll: () => { meshQueue.clearAll(); },
	},
	{
		hasFront: () => particleQueue.sizeFront() > 0,
		hasBack: () => particleQueue.sizeBack() > 0,
		swap: () => { particleQueue.swap(); },
		clearBack: () => { particleQueue.clearBack(); },
		clearAll: () => { particleQueue.clearAll(); },
	},
];

export function prepareCompletedRenderQueues(): void {
	for (let index = 0; index < renderQueueLifecycles.length; index += 1) {
		renderQueueLifecycles[index].swap();
	}
	prepareHeldRenderQueues();
}

function hasCommittedFrontQueueContent(): boolean {
	for (let index = 0; index < renderQueueLifecycles.length; index += 1) {
		if (renderQueueLifecycles[index].hasFront()) return true;
	}
	return false;
}

export function preparePartialRenderQueues(): void {
	activeQueueSource = hasCommittedFrontQueueContent()
		? 'front'
		: (hasPendingBackQueueContent() ? 'back' : 'front');
}

export function prepareOverlayRenderQueues(): void {
	activeQueueSource = 'back';
}

export function prepareHeldRenderQueues(): void {
	activeQueueSource = 'front';
}

export function hasPendingBackQueueContent(): boolean {
	for (let index = 0; index < renderQueueLifecycles.length; index += 1) {
		if (renderQueueLifecycles[index].hasBack()) return true;
	}
	return false;
}

export function clearBackQueues(): void {
	for (let index = 0; index < renderQueueLifecycles.length; index += 1) {
		renderQueueLifecycles[index].clearBack();
	}
	prepareHeldRenderQueues();
}

export function clearAllQueues(): void {
	for (let index = 0; index < renderQueueLifecycles.length; index += 1) {
		renderQueueLifecycles[index].clearAll();
	}
	prepareHeldRenderQueues();
}

export function beginMeshQueue(): number {
	return activeQueueSource === 'back' ? meshQueue.sizeBack() : meshQueue.sizeFront();
}

export function forEachMeshQueue(fn: (item: MeshRenderSubmission, index: number) => void): void {
	if (activeQueueSource === 'back') {
		meshQueue.forEachBack(fn);
		return;
	}
	meshQueue.forEachFront(fn);
}

export function meshQueueBackSize(): number {
	return meshQueue.sizeBack();
}

export function beginParticleQueue(): number {
	return activeQueueSource === 'back' ? particleQueue.sizeBack() : particleQueue.sizeFront();
}

export function forEachParticleQueue(fn: (item: ParticleRenderSubmission, index: number) => void): void {
	if (activeQueueSource === 'back') {
		particleQueue.forEachBack(fn);
		return;
	}
	particleQueue.forEachFront(fn);
}

export let _skyTint: [number, number, number] = [1, 1, 1];
export let _skyExposure = 1.0;
